runtime/src/serde_snapshot.rs

```diff
@@ -954,3 +984,32 @@ where
         ReconstructedAccountsDbInfo { accounts_data_len },
     ))
 }
+
+// Rename `src` to `dest` only if `dest` doesn't already exist.
+#[cfg(target_os = "linux")]
+fn rename_no_replace(src: &CStr, dest: &CStr) -> io::Result<()> {
+    let ret = unsafe {
+        libc::renameat2(
+            libc::AT_FDCWD,
+            src.as_ptr() as *const _,
+            libc::AT_FDCWD,
+            dest.as_ptr() as *const _,
+            libc::RENAME_NOREPLACE,
+        )
+    };
+    if ret == -1 {
+        return Err(io::Error::last_os_error());
```

Comment:
could this return the wrong error if we have multiple os errors close in time together?

Reply:
errno is thread local these days (these days being: the last 20 years), so no. But good observation: once upon a time it wasn't, so multi threaded IO was effectively impossible.